compiler/src/dotty/tools/dotc/typer/Typer.scala

```diff
@@ -2770,4 +2771,9 @@ class Typer extends Namer
       case _ =>
     }
   }
+
+  private def checkStatementPurity(tree: tpd.Tree)(original: untpd.Tree, exprOwner: Symbol)(implicit ctx: Context): Unit = {
+    if (!ctx.isAfterTyper && isPureExpr(tree) && !tree.tpe.isRef(defn.UnitClass) && !isSelfOrSuperConstrCall(tree))
```

Comment:
In this condition, I removed all but the second and the test still passes. I also checked `isSimplyPure` which is stronger, instead of `isPureExpr`, and still passes. 

LGTM! Can you provide a few more tests that exercise the behavior of the other conditions as well?

Reply:
Why only restrict warnings to SimplyPure if we can emit the warning for Pure as well?

Reply:
Ah yes, I agree. I just wanted to point out that there was a test case missing ;-)